stdlib/type.mli

```diff
@@ -35,3 +35,78 @@ type (_, _) eq = Equal: ('a, 'a) eq
 
     At runtime, this function simply returns its second argument unchanged.
 *)
+
+(** {1:identifiers Type identifiers} *)
+
+(** Type identifiers.
+
+    Type identifiers are values that can be tested for
+    {{!Id.equal}equality} to prove they represent the same type. Note
+    that unequal identifiers does not imply unequal types: the same
+    type can be represented by multiple identifiers.
+
+    See an {{!Id.example}example} of use. *)
+module Id : sig
+
+  (** {1:ids Type identifiers} *)
+
+  type 'a t
+  (** The type for type identifiers. ['a] is the represented type. *)
+
+  val make : unit -> 'a t
+  (** [make ()] is a new type identifier. *)
+
+  val equal : 'a t -> 'b t -> ('a, 'b) eq option
+  (** [equal id0 id1] is [Some Equal] if identifier [id0] is equal to [id1]
+      and [None] otherwise. *)
+
+  (** {1:example Example}
+
+      The following shows how type identifiers can be used to implement
+      an heterogeneous key-value dictionary.
+{[
+(** Heterogeneous dictionaries. *)
+module Dict : sig
+  type t
+  (** The type for dictionaries. *)
+
+  type 'a key
+  (** The type for keys binding values of type ['a]. *)
+
+  val key : unit -> 'a key
+  (** [key ()] is a new dictionary key. *)
+
+  val empty : t
+  (** [empty] is the empty dictionary. *)
+
+  val add : 'a key -> 'a -> t -> t
+  (** [add k v d] is [d] with [k] bound to [v]. *)
+
+  val remove : 'a key -> t -> t
+  (** [remove k d] is [d] with the last binding of [k] removed. *)
+
+  val find : 'a key -> t -> 'a option
+  (** [find k d] is the binding of [k] in [d], if any. *)
+end = struct
+  type 'a key = { tid : 'a Type.Id.t; exist : exist_key }
```

Comment:
I think we can simplify the example (avoid the unrelated issues of recursive values and physical equality) with this type:

```suggestion
  type 'a key = 'a Type.Id.t
```

(Wrap the key in `K` when calling list functions and use `assoc` instead of `assq`. This is irrelevant to the example, but this won't even allocate if `K` is `[@@unboxed]` .)

Reply:
I'm uneasy with calling polymorphic comparison on first class modules. What do other people think ?

Reply:
And as mentioned above I think the example is more interesting that way. If people think like @wiktorkuchta, speak up.